pkg/attributes/attributes.go

```diff
@@ -0,0 +1,448 @@
+package attributes
+
+import (
+	"encoding/json"
+	"errors"
+	"strconv"
+
+	apiext "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+)
+
+// Attributes provides a way to add a map of arbitrary YAML/JSON
+// objects.
+// +kubebuilder:validation:Type=object
+// +kubebuilder:validation:XPreserveUnknownFields
+type Attributes map[string]apiext.JSON
+
+// MarshalJSON implements custom JSON marshaling
+// to support free-form attributes
+func (attributes Attributes) MarshalJSON() ([]byte, error) {
+	return json.Marshal(map[string]apiext.JSON(attributes))
+}
+
+// UnmarshalJSON implements custom JSON unmarshaling
+// to support free-form attributes
+func (attributes *Attributes) UnmarshalJSON(data []byte) error {
+	return json.Unmarshal(data, (*map[string]apiext.JSON)(attributes))
+}
+
+// Exists returns `true` if the attribute with the given key
+// exists in the attributes map.
+func (attributes Attributes) Exists(key string) bool {
+	_, exists := attributes[key]
+	return exists
+}
+
+type convertPrimitiveFunc func(attributes Attributes, key string, attributeType string) (interface{}, error)
+
+func (attributes Attributes) getPrimitive(key string, zeroValue interface{}, resultType string, convert convertPrimitiveFunc, errorHolder ...*error) interface{} {
+	var err error
+	if attribute, exists := attributes[key]; exists {
+		var result interface{}
+		switch resultType {
+		case "string":
+			primitiveResult := new(string)
+			err = json.Unmarshal(attribute.Raw, primitiveResult)
+			result = *primitiveResult
+		case "boolean":
+			primitiveResult := new(bool)
+			err = json.Unmarshal(attribute.Raw, primitiveResult)
+			result = *primitiveResult
+		case "number":
+			primitiveResult := new(float64)
+			err = json.Unmarshal(attribute.Raw, primitiveResult)
+			result = *primitiveResult
+		}
+		if err == nil {
+			return result
+		}
+
+		switch typeError := err.(type) {
+		case *json.UnmarshalTypeError:
+			convertedValue, retryError := convert(attributes, key, typeError.Value)
+			if retryError == nil && convertedValue != nil {
+				return convertedValue
+			}
+		}
+	} else {
+		err = errors.New("Attribute with key '" + key + "' does not exist")
+	}
+	if len(errorHolder) > 0 && errorHolder[0] != nil {
+		*errorHolder[0] = err
+	}
```

Comment:
I'm not clear on what we're doing here -- why are we passing in a variadic error if we only ever set the first element?

Reply:
As far as I understand how optional parameters could be achieved in go. So, we could call specify `GetString()` instead of `GetString(nil)`. But it allows us to call `GetString(err, err, err, err)` and the only first is used.

I'm OK with both cases with their pros:
- `GetString()`: which is kind of syntax sugar;
- `GetString(nil)`: is a bit more explicit and suggests that error is ignored - in case you know parameter type;

Reply:
You're right @sleshchenko. I just used variadic parameter to provide an optional parameter.
However in this form this API also allows, in the future, passing an array of errors and possibly getting as many errors as the provide variadic list allows, instead of only returning the last error encountered.

Reply:
>  passing an array of errors and possibly getting as many errors as the provide variadic list allows, instead of only returning the last error encountered.

Has client initialize errors instances? I mean the client is not able to consume the items provided in the array, is it?
If not, then how you see it in the future? Like client initializes 100 errors and some of them are modified?

Reply:
@sleshchenko I'm OK to change it to `GetString(key string, errorHolder *error)`,
and enforce the client to explicitly pass a `nil` error pointer in the second argument to clearly state it ignores the error.

@maysunfaisal @yangcao77 @amisevsk any preference between the use of an optional parameter through the use of variadics or a single error pointer ?

Reply:
I think an optional parameter would be nice to have, but also allows us to implicitly skip error checking with no way to easily notice it in code. Without knowing the API we're using in detail, it might be hard to notice e.g. 
```go
  myString := attributes.GetString("myKey")
```
is discarding an error silently.

We're kind of departing general Go paradigms for error handling (which I'm not a fan of anyways), so it's hard to reason about what makes sense here. I'm generally fine with whatever we choose, but it might be confusing for (eventual) newcomers.

Reply:
So maybe @sleshchenko proposal of using `GetString("aString", nil)` is the best here ?This would still allow using `GetString` in expressions (by avoiding multiple value returns) while avoiding silent error discarding ?

Reply:
Done in commit https://github.com/devfile/api/pull/214/commits/2c5bf7950b093c7edbc8c77b3a6d2463e178ddc5